cmd/fastqReplaceSeq/fastqReplaceSeq.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fastq"
+	"github.com/vertgenlab/gonomics/fileio"
+)
+
+type FindReplaceSeq struct {
+	find    []dna.Base
+	replace []dna.Base
+}
+
+// readFindReplaceDNA converts a delim separated findReplace file to a []FindReplaceSeqs
+func readFindReplaceDNA(filename string, delim string) []FindReplaceSeq {
+	var in *fileio.EasyReader
+	var findReplacePairs []FindReplaceSeq
+	var currPair FindReplaceSeq
+	var seqs []string
+	var done bool
+	var line string
+	var err error
+
+	in = fileio.EasyOpen(filename)
+	for line, done = fileio.EasyNextLine(in); !done; line, done = fileio.EasyNextLine(in) {
+		seqs = strings.Split(line, delim)
+
+		if len(seqs) != 2 {
+			log.Fatalf("Error: the following line:\n\"%s\"\ndoes not give two substrings when split with \"%s\"", line, delim)
+		} else if len(seqs[0]) != len(seqs[1]) {
+			log.Fatalf("Error: find sequence must be same length as replace sequence.\n")
+		}
+
+		currPair.find = dna.StringToBases(seqs[0])
+		currPair.replace = dna.StringToBases(seqs[1])
+		findReplacePairs = append(findReplacePairs, currPair)
+	}
+	err = in.Close()
+	exception.PanicOnErr(err)
+
+	return findReplacePairs
+}
+
+// findSeq loops through a query sequence starting at the beginning and compares it to a find sequence, base by base
+// outputs true if find sequence is found in query sequence
+func findSeq(seq []dna.Base, find []dna.Base, ignoreCase bool) bool {
+	if len(find) > len(seq) {
+		log.Fatalf("Error: Length of find sequence must be less then or equal to length of query sequence.")
+	}
+	for currBase := range find {
+		if dna.CompareBases(seq[currBase], find[currBase], ignoreCase) != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// replacePrefix replaces bases in a query sequence with a replace sequence, starting from the beginning of the query sequence
+func replacePrefix(seq []dna.Base, replace []dna.Base) []dna.Base {
+	copy(seq, replace[0:])
+	return seq
+}
+
+type Settings struct {
+	inFile              string
+	outFile             string
+	findReplaceFile     string
+	findReplaceDelim    string
+	ignoreCase          bool
+	replacedRecordsOnly bool
+}
+
+func usage() {
+	fmt.Print(
+		"fastqReplaceSeq - finds a sequence in a fastq file (starting from the beginning of the line) and replaces it with a new sequence of the same length. \n" +
+			"Usage:\n" +
+			"fastqReplaceSeq input.fastq findReplaceFile output.fastq\n" +
+			"options:\n")
+	flag.PrintDefaults()
+}
+
+func fastqReplaceSeq(s Settings) {
+	var findReplacePairs []FindReplaceSeq
+	var currPair FindReplaceSeq
+	var find []dna.Base
+	var replace []dna.Base
+	var replaceCounter int = 0
+	fq := fastq.GoReadToChan(s.inFile)
+	out := fileio.EasyCreate(s.outFile)
+	var err error
+
+	findReplacePairs = readFindReplaceDNA(s.findReplaceFile, s.findReplaceDelim)
+
+	for currRecord := range fq {
+		found := false
+		for _, currPair = range findReplacePairs {
```

Comment:
Instead of having to loop through all the find and replace pairs for every fastq entry, what if you store the find and replace in a map as strings. I am thinking with the find as key and replace as the value. You can then query the map with the fastq sequence (as a string), which will tell you if the sequence needs to be replaced without the need to loop. I think that might be much faster